### ISSUE TITLE
Get pino and jest to agree on console verbosity

### ISSUE THREAD

**Issue by bickelj** [PhilanthropyDataCommons/service#59]
See comment in PR #55 at https://github.com/PhilanthropyDataCommons/service/pull/55#discussion_r927000538

An example possibility is at https://github.com/chriszs/tv-kitchen/commit/ff61ea8611d092740e38a020b06d7aff4cd979f1

**Comment by bickelj** [comment]
It seems backwards to have the production code look to see if it's in a test environment first and then change the level based on that. Perhaps there is a way to have the test environment set the level instead.

**Comment by bickelj** [comment]
Here's an approach:
1. Drive the log level with env var `PINO_LOG_LEVEL` or `info` if it is not set.
2. In the jest configuration, when --silent is set and `PINO_LOG_LEVEL` is not set, set `PINO_LOG_LEVEL` to `silent`.

I think the above will mean:
* When `PINO_LOG_LEVEL` is set in prod or test, that level takes effect.
* When `PINO_LOG_LEVEL` is not set in prod, `info` takes effect.
* When `PINO_LOG_LEVEL` is not set in test, and also --silent is set, then the test harness/framework will set `PINO_LOG_LEVEL` to `silent` and it will therefore take effect.

**Comment by bickelj** [comment]
The problem is finding where in jest we can hook in to have access to the resolved/merged configuration object. I have yet to find it in the docs.

**Comment by bickelj** [comment]
The closest thing I can find is our existing `setupEnv.ts` which is akin to the above linked file listed in `setupFiles`, but then there is the problem of importing the logger after detecting whether jest has set the console to silent. I think the root issue again is figuring out how to programmatically read the `jest` configuration after it has been resolved. I cannot see any way to do it despite other folks wondering the same. No answers at https://stackoverflow.com/questions/65698821/how-to-read-jest-config-values-within-test and I cannot find anything in the docs or source code indicating they would be exposed. That doesn't mean it isn't there, though.

**Comment by bickelj** [comment]
Perhaps it is the `globalSetup` that I seek: https://jestjs.io/docs/configuration#globalsetup-string

**Comment by bickelj** [comment]
The use of `globalSetup` seems to work.

**Comment by bickelj** [comment]
Unless I am missing something obvious (I probably am), I can choose between the linter complaining and jest complaining.

With `module.exports = function ...` I get this from the linter (but jest complaineth not):

> Cannot use import declarations in modules that export using CommonJS (module.exports = 'foo' or exports.bar = 'hi')  import/no-import-module-exports

With `export function ...` I get this from jest (but the linter complaineth not):

> globalSetup file must export a function

**Comment by slifty** [comment]
This is a case where Jest is relying on a specific export pattern which we have chose NOT to use in our code -- I think it's a valid candidate for a lint modification.

You shouldn't use modules.export (that is old school node syntax) but you SHOULD use export default instead of exporting a named function (since jest is expecting it to be default).

So add this at the top of the file

```
// Jest expects a single default function to be exported from this file.
// eslint-disable import/no-default-export
```

As a side note: I would recommend we avoid using the `function` keyword outside of class methods (arrow functions are a more recent addition to javascript and have some subtle perks that make them slightly safer; it's also just nice to have consistency and they're shorter to write) -- here's a longer thread from another OTS project where this was hotly discussed: https://github.com/PermanentOrg/permanent-sdk/pull/14#issuecomment-1062256449

So instead of:

```
export default function myFunc() { ... }
```

You might do something like 

```
export default myFunc = () => { ... }
```

(I know we have used `function` already in a few places on `main`, but when we have a moment to breathe we may want to change those.  Unfortunately this isn't something that eslint knows how to lint for just yet...)

**Comment by bickelj** [comment]
> As a side note: I would recommend we avoid using the `function` keyword outside of class methods

I think I might have started with an arrow function but then while going through the "is this it? nope" iterations I found the solution I was looking for, unrelated to that. Good call, I have made it an (anonymous) arrow function.

> So add this at the top of the file...

Also good call, done!

**Comment by bickelj** [comment]
It makes sense here, considering this is about pino and jest working well together with regard to logging. I personally have been using pino-pretty off and on when running locally and setting the log level to have pino show up. For example:
```shell
PINO_LOG_LEVEL=debug npm run test | npx pino-pretty
```
But also if I'm not interested in the pino output, run it plain:
```shell
npm run test
```
The caller has options. I suppose these should be documented more clearly, though. I can add some examples to the `README.md`, I suppose.

On another note, it doesn't look like I said it before, but examining `console.constructor.name` (the solution shown in the original post) did not seem to work for me, that is why I kept trying to find a more explicit or direct or congruent solution.

**Comment by bickelj** [comment]
While making this agree with PR #55 on the name of the env var, I thought I would add something to the README. When I did that, I noticed our `.env` file and `.env.test` files which should be loaded with `dotenv.config`. However, when setting `LOG_LEVEL` in the `.env` file, it did not have the desired effect during tests, so I wonder if that code needs to go into the `globalSetup` rather than `setupFiles` in order to take effect. Anyway, this can be resolved elsewhere I guess.

**Comment by bickelj** [comment]
OK, since I found it, trying it with the dotenv in `globalSetup` really quickly.

**Comment by bickelj** [comment]
Yeah, that behaves better to set it in the `globalSetup`. Doing that instead.

**Comment by bickelj** [comment]
Works great in test, but in prod, these are out of order, dependency wise:
```TypeScript

import { logger } from './logger';

dotenv.config();
```

**Comment by slifty** [comment]
This might sound terrible to say out loud BUT -- it's appropriate for us to call `dotenv.config()` in `logger.js`

So here's what we need to do.

`test/setupEnv.ts` should be updated to specify the `DOTENV_CONFIG_PATH` env var. <-- this is a bug in itself, I can fix that in another PR right now super fast.

**Comment by bickelj** [comment]
My solution for test was to make `globalSetup` look like this, which works great:
```TypeScript
// Jest expects a single default function to be exported from this file.
/* eslint-disable import/no-default-export */
import dotenv from 'dotenv';
import type { Config } from 'jest';

dotenv.config({ path: '.env.test' });

export default (globalConfig: Config, projectConfig: Config): void => {
  if ((process.env.LOG_LEVEL === undefined || process.env.LOG_LEVEL === '')
      && (globalConfig.silent === true || projectConfig.silent === true)) {
    process.env.LOG_LEVEL = 'silent';
  }
};
```
As for prod, yeah, that sounds really ugly, I'd rather create a separate production `aSetupEnv.ts` that calls `dotenv.config()` and that way it happens first (the 'a' being first in the alphabet for import order?).

**Comment by bickelj** [comment]
I think `test/setupEnv.ts` happens much later, perhaps per-test or per-suite, after the `globalSetup`, so it is too late by then. It seems appropriate to do the `dotenv` in `globalSetup` to me.

**Comment by slifty** [comment]
Agreed; also I think that the expected dotenv use is to ONLY run it once but to run it at entrypoint before config files are invoked.  I think you wound up there, that's a much better approach.

I don't think we need to think about a `setupEnv.ts` outside of tests -- It also sounds like we can delete the `setupEnv` yes?

**Comment by bickelj** [comment]
Yes, if the `dotenv.config()` call happens in `globalSetup.ts` then we would delete `setupEnv.ts`.

For now, I will go with the call to `dotenv.config()` in `logger.ts` just to get over the hurdle and we can figure something out later if needed.

**Comment by bickelj** [comment]
OK, this is stickier than at first glance. So I think don't wait for this to rebase the other PR.

**Comment by bickelj** [comment]
I think `DOTENV_CONFIG_PATH` is a one-shot thing or must be set before loading the `dotenv` module. Setting it in code (at least after importing `dotenv` does not seem to take any effect.

**Comment by bickelj** [comment]
OK, I think #60 now has a congruent setup, finally!

**Comment by bickelj** [comment]
Resolved in #60
